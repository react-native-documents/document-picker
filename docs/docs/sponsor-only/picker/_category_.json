{
  "label": "Document Picker",
  "position": 20,
  "collapsed": false,
  "link": {
    "type": "generated-index"
  }
}
