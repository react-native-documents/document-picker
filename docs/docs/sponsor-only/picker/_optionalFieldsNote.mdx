:::note
Many of the fields are nullable because the file metadata might not be available in some cases. While it's unlikely, it can happen - especially on Android - if a user picks a file from a [Document Provider](https://developer.android.com/guide/topics/providers/create-document-provider) that doesn't make the information available.
:::
