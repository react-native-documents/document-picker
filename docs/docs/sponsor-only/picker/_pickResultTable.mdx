import OptionalFieldsNote from './_optionalFieldsNote.mdx'

<OptionalFieldsNote />

Each picked file is represented by an object with the following properties:

| Name                     | Type                 | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           |
| :----------------------- | :------------------- | :-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `uri`                    | `string`             | The URI of the picked file. Note that it is encoded, so you might need to [decode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/decodeURI) it for further processing.                                                                                                                                                                                                                                                                                                             |
| `name`                   | `string` \| `null`   | The name of the picked file, including the extension. It's very unlikely that it'd be `null` but in theory, it can happen.                                                                                                                                                                                                                                                                                                                                                                                            |
| `size`                   | `number` \| `null`   | The size of the picked file in bytes.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 |
| `type`                   | `string` \| `null`   | The MIME type of the picked file.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
| `hasRequestedType`       | `boolean`            | Android: Some [Document Providers](https://developer.android.com/guide/topics/providers/content-provider-basics) on Android (especially those popular in Asia, it seems) do not respect the request for [limiting selectable file types](./limiting-selectable-files.md). `hasRequestedType` will be false if the user picked a file that does not match one of the requested types. You need to do your own post-processing and display an error to the user if this is important to your app. Always `true` on iOS. |
| `error`                  | `string` \| `null`   | Error in case the file metadata could not be obtained.                                                                                                                                                                                                                                                                                                                                                                                                                                                                |
| `isVirtual`              | `boolean` \| `null`  | Android: Whether the file is a virtual file (such as Google docs or sheets). Will be `null` on pre-Android 7.0 devices. On iOS, it's always `false`.                                                                                                                                                                                                                                                                                                                                                                  |
| `convertibleToMimeTypes` | `string`[] \| `null` | Android: The target types to which a virtual file can be converted. Useful for `keepLocalCopy`. This field is only specified if `isVirtual` is true, and only on Android 7.0+. Always `null` on iOS.                                                                                                                                                                                                                                                                                                                  |
| `nativeType`             | `string` \| `null`   | The "native" type of the picked file: on Android, this is the MIME type. On iOS, it is the UTType identifier.                                                                                                                                                                                                                                                                                                                                                                                                         |
