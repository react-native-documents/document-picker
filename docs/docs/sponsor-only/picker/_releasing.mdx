### Releasing Long Term Access

This is an Android-only feature. When you no longer need access to the file or location, you should release the long-term access by calling `releaseLongTermAccess`. Calling this on iOS will resolve.

See [Android documentation](<https://developer.android.com/reference/android/content/ContentResolver#releasePersistableUriPermission(android.net.Uri,%20int)>) for more information.

### Releasing (stopping) Secure Access

This is an iOS-only feature. When you no longer need access to the file or location, you should release the secure access by calling `releaseSecureAccess`. Calling this on Android will resolve.

See [iOS documentation](https://developer.apple.com/documentation/foundation/nsurl/1413736-stopaccessingsecurityscopedresou) for more information.
