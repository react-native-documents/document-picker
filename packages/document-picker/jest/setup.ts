import type { Spec } from '../src/spec/NativeDocumentPicker'

import {
  DirectoryPickerResponse,
  DocumentPickerResponse,
  IsKnownTypeResponse,
  KeepLocalCopyOptions,
  LocalCopyResponse,
} from '../src'

export const mockDocPickerResponse: DocumentPickerResponse = {
  uri: 'file:///mock/uri/mockName.pdf',
  name: 'mockName.pdf',
  type: 'application/pdf',
  nativeType: 'com.adobe.pdf',
  size: 1234,
  error: null,
  isVirtual: false,
  convertibleToMimeTypes: null,
  hasRequestedType: true,
}

function mockFactory() {
  const mockNativeModule: Spec = Object.freeze({
    isKnownType: jest.fn().mockReturnValue({
      isKnown: false,
      UTType: null,
      preferredFilenameExtension: null,
      mimeType: null,
    } satisfies IsKnownTypeResponse),
    keepLocalCopy: jest.fn().mockImplementation(async (options: KeepLocalCopyOptions) => {
      const mockCopyResults: LocalCopyResponse[] = options.files.map(({ uri, fileName }) => ({
        sourceUri: uri,
        localUri: `file:///mock/uri/${fileName}`,
        status: 'success',
      }))
      return mockCopyResults
    }),
    pick: jest.fn().mockResolvedValue([mockDocPickerResponse] satisfies DocumentPickerResponse[]),
    pickDirectory: jest.fn().mockResolvedValue({
      uri: 'file:///mock/uri/mockDirectory',
    } satisfies DirectoryPickerResponse),
    saveDocument: jest
      .fn()
      .mockResolvedValue([mockDocPickerResponse] satisfies DocumentPickerResponse[]),
    writeDocuments: jest
      .fn()
      .mockResolvedValue([mockDocPickerResponse] satisfies DocumentPickerResponse[]),
    releaseSecureAccess: jest.fn().mockResolvedValue(null),
    releaseLongTermAccess: jest.fn().mockResolvedValue([]),
  })
  return {
    NativeDocumentPicker: mockNativeModule,
  }
}

jest.mock('../src/spec/NativeDocumentPicker', () => mockFactory())
// the following are for jest testing outside of the library, where the paths are different
// alternative is to use moduleNameMapper in user space
const mockModulePaths = [
  '../../../lib/commonjs/spec/NativeDocumentPicker',
  '../../../lib/module/spec/NativeDocumentPicker',
]
mockModulePaths.forEach((path) => {
  try {
    require.resolve(path)
    jest.mock(path, () => mockFactory())
  } catch (error: any) {
    if ('code' in error && error.code === 'MODULE_NOT_FOUND') {
      if (!process.env.SILENCE_MOCK_NOT_FOUND) {
        console.warn(`Unable to resolve ${path}`)
      }
    } else {
      throw error
    }
  }
})
