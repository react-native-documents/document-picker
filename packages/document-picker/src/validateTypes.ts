import type { DocumentPickerResponse } from './types'

export const safeValidate = (
  requestedTypes: string[],
  singlePickResult: Pick<DocumentPickerResponse, 'nativeType'>,
) => {
  const { nativeType } = singlePickResult
  return requestedTypes.some((it) => nativeTypeMatches(it, nativeType))
}

const nativeTypeMatches = (requiredType: string, actualType: string | null) => {
  if (requiredType === actualType || requiredType === '*/*') {
    return true
  }
  if (actualType === null) {
    return false
  }
  if (requiredType.endsWith('/*')) {
    const requiredTypeWithoutStar = requiredType.slice(0, -2)
    if (actualType.startsWith(requiredTypeWithoutStar)) {
      return true
    }
  }

  return false
}
