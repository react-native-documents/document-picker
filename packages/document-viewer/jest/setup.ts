type ExportedModuleType = typeof import('../src/index')

const mockModule: ExportedModuleType = Object.freeze({
  viewDocument: jest.fn().mockResolvedValue(null),
})

jest.mock('@react-native-documents/viewer', () => {
  return mockModule
})
