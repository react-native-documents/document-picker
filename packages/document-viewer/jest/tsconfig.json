{
  "extends": "@tsconfig/node18/tsconfig.json",
  "compilerOptions": {
    "outDir": "./build",
    "jsx": "react",
    "lib": ["es2022"],
    "declaration": false
  },
  "include": ["./setup.ts"],
}
