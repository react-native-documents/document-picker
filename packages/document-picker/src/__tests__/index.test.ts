import '../../jest/setup' // set up the mock, same effect as setupFiles
import { expect, it } from '@jest/globals'

// this import path not is mocked, we're mocking on the native module boundary
import * as RNDP from '@react-native-documents/picker'
import { safeValidate } from '../validateTypes'

describe('DocumentPicker mock should', () => {
  describe('safeValidate', () => {
    it.each([
      // [requestedTypes, type, expected]
      [['*/*'], null, true],
      [['*/*'], 'image/jpeg', true],
      [['image/jpeg'], 'image/jpeg', true],
      [['image/jpeg', 'image/png'], 'image/png', true],
      [['application/pdf'], 'image/jpeg', false],
      [['image/*'], 'image/gif', true],
      [['image/*'], 'video/mp4', false],
      [['application/*'], 'application/json', true],
      [['text/*', 'image/*'], 'text/html', true],
      [['text/*', 'image/*'], 'application/json', false],
      [['image/jpeg'], null, false], // Testing null type in singlePickResult
      [['image/*'], null, false],
    ])(
      'correctly validates requestedTypes %s with type %s as %s',
      (requestedTypes, nativeType, expected) => {
        const singlePickResult = { nativeType }
        const result = safeValidate(requestedTypes, singlePickResult)
        expect(result).toBe(expected)
      },
    )
  })

  it('have the expected methods', () => {
    expect(RNDP).toMatchSnapshot()
  })

  it('allow picking a mock document', async () => {
    const response = await RNDP.pick()
    expect(response).toStrictEqual([
      {
        convertibleToMimeTypes: null,
        error: null,
        hasRequestedType: true,
        isVirtual: false,
        name: 'mockName.pdf',
        nativeType: 'com.adobe.pdf',
        size: 1234,
        type: 'application/pdf',
        uri: 'file:///mock/uri/mockName.pdf',
      },
    ])
  })
  //.. todo can test more mock methods
})
