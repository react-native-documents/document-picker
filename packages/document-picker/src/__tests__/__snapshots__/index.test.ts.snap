// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentPicker mock should have the expected methods 1`] = `
{
  "errorCodes": {
    "IN_PROGRESS": "ASYNC_OP_IN_PROGRESS",
    "OPERATION_CANCELED": "OPERATION_CANCELED",
    "UNABLE_TO_OPEN_FILE_TYPE": "UNABLE_TO_OPEN_FILE_TYPE",
  },
  "isErrorWithCode": [Function],
  "isKnownType": [Function],
  "keepLocalCopy": [Function],
  "pick": [Function],
  "pickDirectory": [Function],
  "releaseLongTermAccess": [Function],
  "releaseSecureAccess": [Function],
  "saveDocuments": [Function],
  "types": {
    "allFiles": "public.item",
    "audio": "public.audio",
    "csv": "public.comma-separated-values-text",
    "doc": "com.microsoft.word.doc",
    "docx": "org.openxmlformats.wordprocessingml.document",
    "images": "public.image",
    "json": "public.json",
    "pdf": "com.adobe.pdf",
    "plainText": "public.plain-text",
    "ppt": "com.microsoft.powerpoint.ppt",
    "pptx": "org.openxmlformats.presentationml.presentation",
    "video": "public.movie",
    "xls": "com.microsoft.excel.xls",
    "xlsx": "org.openxmlformats.spreadsheetml.sheet",
    "zip": "public.zip-archive",
  },
}
`;
